src/coreclr/vm/appdomain.cpp

```diff
@@ -5115,6 +5084,17 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
                 COMPlusThrowHR(COR_E_INVALIDOPERATION, IDS_HOST_ASSEMBLY_RESOLVER_DYNAMICALLY_EMITTED_ASSEMBLIES_UNSUPPORTED, name);
             }
 
+            // For collectible assemblies, ensure that the parent loader allocator keeps the assembly's loader allocator
+            // alive for all its lifetime.
+            if (pDomainAssembly->IsCollectible())
```

Comment:
Hmm, but we need to create the link even in that case (and also in case of non-default non-collectible context) or rather prevent that from happening by throwing an exception, otherwise the ALC we've resolved from would get collected and we would end up with the issue as before. I have thought we do prevent that already, but local testing has revealed we don't.

Reply:
Oh, hm, yeah, we are definitely missing that case then, since `EnsureReference` also won't add the reference for non-collectible.

Looks like we error out on collectible assemblies resolved by `AppDomain.AssemblyResolve`.
https://github.com/dotnet/runtime/blob/eb9dd67ce9fc64c468cf5c795fda7057310c22fc/src/coreclr/vm/appdomain.cpp#L4640-L4643
But we don't check the results of other extension points.

Reply:
@elinor-fung I've added the exception throwing for NULL parent loader allocator.